sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc

```diff
@@ -0,0 +1,229 @@
+|====
+
+== Modifications to the SPIR-V Specification, Version 1.2
+
+=== Capabilities
+
+Modify Section 3.31, Capability, adding rows to the Capability table:
+
+[cols="1,10,4,4",options="header"]
+|====
+2+^| Capability ^| Implicitly Declares ^| Enabled by Extension
+
+| {capability_token}
+| *VariableLengthArrayINTEL*
+|
+| *SPV_INTEL_variable_length_array*
+
+|====
+
+=== Instructions
+
+Modify Section 3.36.8. Memory Instructions, adding to the end of the list of
+instructions:
+
+[cols="1,1,3*3",width="100%"]
+|=====
+4+|[[OpVariableLengthArrayINTEL]]*OpVariableLengthArrayINTEL* +
+ +
+Allocate a runtime-sized array in the function memory, resulting in a pointer to
+it, which can be used with *OpLoad* and *OpStore*. The allocated memory is
+uninitialized. Loading from uninitialized memory causes undefined behavior. Size
+of memory to be allocated computed as the size of 'Type' operand of the
+'Result Type' multiplied by the 'Length'.
+ +
+
+'Result Type' must be *OpTypePointer* with *Function* storage class.
+Its 'Type' operand is the type of array elements. The type of array elements
+must be a concrete type.
+ +
+
+'Length' is the number of elements in the array. It must have a scalar integer
+type. 'Length' is treated as unsigned, and if its value is 0, no memory will be
+allocated. If there is insufficient memory space for the allocation the behavior
+is undefined.
+ +
+1+|Capability: +
+*VariableLengthArrayINTEL*
+| 4 | {OpVariableLengthArrayINTEL_token}
+| '<id>' 'Result Type'
+| '<id>' 'Result'
+| '<id>' 'Length'
+|=====
+
+
+[cols="1,1,2*3",width="100%"]
+|=====
+3+|[[OpSaveMemoryINTEL]]*OpSaveMemoryINTEL* +
+ +
+Save the current state of the function memory. Returns a pointer that should be
+passed to <<OpRestoreMemoryINTEL,*OpRestoreMemoryINTEL*>>.
+ +
+
+'Result Type' must be *OpTypePointer* with *Function* storage class.
+ +
+1+|Capability: +
+*VariableLengthArrayINTEL*
+| 3 | {OpSaveMemoryINTEL_token}
+| '<id>' 'Result Type'
+| '<id>' 'Result' 
+|=====
+
+
+[cols="1,1,1*3",width="100%"]
+|=====
+2+|[[OpRestoreMemoryINTEL]]*OpRestoreMemoryINTEL* +
+ +
+Restore the function memory to the state it was in when the
+<<OpSaveMemoryINTEL,*OpSaveMemoryINTEL*>> was executed. Effectivelly, this
+instruction deallocates the function memory allocated by
+<<VariableLengthArrayINTEL,*VariableLengthArrayINTEL*>>.
+ +
+
+'Ptr' is a pointer value returned by <<OpSaveMemoryINTEL,*OpSaveMemoryINTEL*>>.
+It must be an *OpTypePointer* with *Function* storage class.
+
+1+|Capability: +
+*VariableLengthArrayINTEL*
+| 2 | {OpRestoreMemoryINTEL_token} | '<id>' 'Ptr'
+|=====
+
+
+== Validation Rules
+
+ - In control flow graph every
+<<VariableLengthArrayINTEL,*VariableLengthArrayINTEL*>> must be dominated by at
+least one <<OpSaveMemoryINTEL,*OpSaveMemoryINTEL*>>.
+
+ - In control flow graph every
+<<VariableLengthArrayINTEL,*VariableLengthArrayINTEL*>> must be postdominated by
+at least one <<OpSaveMemoryINTEL,*OpSaveMemoryINTEL*>>.
```

Comment:
Should it be restore instead of save here?

Reply:
Good catch! Thanks.